### ISSUE TITLE
BatchNorm on variable-length sequences or batches

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#70529]
### 🚀 The feature, motivation and pitch

BatchNorm1d is often used on variable-length sequences. As I understand, that would be always incorrect currently, as it ignores the sequence lengths and thus also includes the padded values for the statistics.

To fix this, there should be an additional argument like `seq_lengths` or `batch_sizes` or so, and it should calculate the statistics only over the valid values.

(Similarly also for 2D and 3D batch norm.)

### Alternatives

The alternative is the current way to just ignore this.

I wonder a bit that this is not discussed or even mentioned at all in the documentation.

### Additional context

I don't really know the influence of this. It probably depends. I see Batch Norm also being used for sequence tasks like speech recognition, machine translation, text-to-speech and others, where the usage is almost always "wrong". I assume not many people have really investigated the effect of this.

If the effect is minor, then ignoring is maybe fine. But I would expect some discussion, and a reference to some study on this.

If the effect is big, then this is a serious problem. Then this is not really just a feature request but a serious issue.

If the effect is unknown, then I'm not sure if just ignoring it is such a good idea.

I have tried to search a bit for any references, discussions, study, or just mentions on this but I failed to find much.

#25910
https://discuss.pytorch.org/t/padding-for-variable-length-sequences-and-batch-norm-layer/21985

Also, TensorFlow seems to ignore it as well.

**Comment by albertz** [comment]
Despite correct masking and using the seq length (num frames) for calculating the statistics of the current mini-batch, also the momentum should take the seq length (num frames) into account. Using `num_batches_tracked` is not enough, it should be `num_frames_tracked`.

**Comment by SeanNaren** [comment]
@albertz reviving from the dead here as you also left a message a while back on the deepspeech.pytorch repo! I think you are correct, most libraries (including mine) do not exclude padding, however I think the assumption made is that similar sequence lengths are grouped together (how true this is for all implementations I cannot tell you).

**Comment by dhruvbird** [comment]
I just ran into this same problem. I have a seq2seq LSTM model that takes variable size inputs, so they are padded for batched training. However, for inference I pass a batch of size 1. This resulted in an ~10% accuracy drop. If I run inference with the same batch size as training (i.e. 128 in my case) then the accuracy is fine.

**Comment by wenjie-p** [comment]
I also wonder how this will influence the final model performance. If we don't exclude the padded values, then the statistics may depend on the value used for padding, i.e. zero-padding or padding with repeating last sample.

**Comment by Abdulrhman-shaheen** [comment]
I’m surprised this issue has received so little attention. I recently encountered a concrete downstream instance of the same behavior in Hugging Face Transformers: https://github.com/huggingface/transformers/issues/47739

While a controlled experiment is needed to quantify the actual impact on models performance, the effects on activations, gradients, and running statistics are easy to be shown (one example for dummy input is in the issue linked above). This should be sufficient to show that padding-aware BatchNorm is needed.

I haven't done much research in the current implementation but I believe two major operations need to change:
1. Exclude padding from the statistics calculation.
1. Restore padded position to zero after applying batch norm.

I would be happy to contribute an implementation if a maintainer is available to discuss the preferred API (whether to have an exclusive `masked_batch_norm` or extend the current implementation) and expected semantics (restoring the padding should be the caller responsibility or should be handled in `batch_norm` itself).
